feat(common): rename underlying `NgFor` class and add a type parameter

chuckjaz opened the pull request:
feat(compiler): generate a type parameters for generic type references
feat(tsc-wrapped): record arity of generic classes

Note, this affects the underlying class and should not affect usage.

DEPRECATION:
- the `NgFor` class is now deprecated. Use `NgForOf<T>` instead.
  IMPORTANT: Only the `NgFor` class is deprecated, not the `ngFor`
  directive. The `*ngFor` and related directives are unaffected by
  this change  as references to the `NgFor` class generated from
  templates will be automatically converted to references to
  `NgForOf<T>` without requiring any template modifications.
- `TrackByFn` is now deprecated. Use `TrackByFunction<T>` instead.

Migration:
- Replace direct references to the `NgFor` class to `NgForOf<any>`.
- Replace references to `TrackByFn` to `TrackByFunction<any>`.

BREAKING CHANGE:
A definition of `Iterable<T>` is now required to correctly compile
Angular applications. Support for `Iterable<T>` is not required at
runtime but a type definition `Iterable<T>` must be available.

`NgFor`, and now `NgForOf<T>`, already supports `Iterable<T>` at
runtime. With this change the type definition is updated to reflect
this support.

Migration:
- add "es2015.iterable.ts" to your tsconfig.json "libs" fields.

**What kind of change does this PR introduce?** (check one with "x")
```
[x] Feature
```

**What is the current behavior?** (You can also link to an open issue here)

`NgFor` uses `NgForRow` for its context which has a `$implicit` type of `any`.  The type of `$implicit` should be the element type of the iterator assigned to the `ngForOf` property.

The class name `NgFor` is incorrect as well as it implies it is associated with the `ngFor` directive instead of the `ngForOf` directive. Calling it `NgFor` seems to imply that if we add syntax to support iterating the fields of an object such as `for...in`, `NgFor` would need to host the code instead of creating a new class.

The TypeScript code generation does not supply the correct number of type arguments to classes with generic parameters.

**What is the new behavior?**

A new class is created, `NgForOf<T>` and the old class is deprecated and transformed into a derived class `class NgFor extends NgForOf<any> {}`. 

The TypeScript emitter was changed to generate the correct number of type parameters for references to generic types.

The TypeScript code generation will generate `NgForOf<any>` for references to this class. However, once #12398 lands, the type parameter would be used to validate references to the context`$implicit`

**Does this PR introduce a breaking change?** (check one with "x")
```
[x] Yes
[ ] No
```

This change requires the project's `tsconfig.json` to include a definition of `Iterable<T>` to in the global scope. This can be accomplished by adding "es2015.iterable" into the `"libs"` field.

vicb commented:
```
feat(compile): generate a type parameters for generic type references
feat(tsc-wrapped): record arity of generic classes
```
Should this be 2 separate commits ?

chuckjaz commented:
Sure. Do you want three commits, one for adding `arity`, one for adding generics and one for renaming `NgFor` etc?

chuckjaz committed:
feat(tsc-wrapped): record arity of generic classes (#14104)

chuckjaz committed:
feat(common): rename underlying `NgFor` class and add a type parameter (#14104)

Note, this affects the underlying class and should not affect usage.

DEPRECATION:
- the `NgFor` class is now deprecated. Use `NgForOf<T>` instead.
  IMPORTANT: Only the `NgFor` class is deprecated, not the `ngFor`
  directive. The `*ngFor` and related directives are unaffected by
  this change  as references to the `NgFor` class generated from
  templates will be automatically converted to references to
  `NgForOf<T>` without requiring any template modifications.
- `TrackByFn` is now deprecated. Use `TrackByFunction<T>` instead.

Migration:
- Replace direct references to the `NgFor` class to `NgForOf<any>`.
- Replace references to `TrackByFn` to `TrackByFunction<any>`.

BREAKING CHANGE:
A definition of `Iterable<T>` is now required to correctly compile
Angular applications. Support for `Iterable<T>` is not required at
runtime but a type definition `Iterable<T>` must be available.

`NgFor`, and now `NgForOf<T>`, already supports `Iterable<T>` at
runtime. With this change the type definition is updated to reflect
this support.

Migration:
- add "es2015.iterable.ts" to your tsconfig.json "libs" fields.

Part of #12398

tbosch commented on modules/@angular/compiler/src/aot/static_symbol_resolver.ts:
```diff
@@ -83,6 +84,15 @@ export class StaticSymbolResolver {
+   * summaries re-export instead of directly to avoid introducing a direct dependency
+   * on an otherwise indirect dependency.
+   *
+   * @param statiSymbol the symbol for which to generate a import symbol
```
type `staticSymbol`

tbosch commented on modules/@angular/compiler/src/aot/static_symbol_resolver.ts:
```diff
@@ -83,6 +84,15 @@ export class StaticSymbolResolver {
+   * getImportAs produces a symbol that can be used to import the given symbol.
+   * The import might be different than the symbol if the symbol is exported from
+   * a library with a summary; in which case we want to import the symbol from the
+   * summaries re-export instead of directly to avoid introducing a direct dependency
```
from the `ngfactory`

tbosch commented on modules/@angular/compiler/src/aot/static_symbol_resolver.ts:
```diff
@@ -134,6 +163,8 @@ export class StaticSymbolResolver {
    * @param declarationFile the absolute path of the file where the symbol is declared
    * @param name the name of the type.
+   * @param members a symbol for a static member of the named type
+   * @param arity the number of generic type parameters or undefined if there are none
```
You don't pass `arity` here as argument...

tbosch commented on modules/@angular/compiler/src/output/output_ast.ts:
```diff
@@ -881,13 +877,12 @@ export function importExpr(id: CompileIdentifierMetadata, typeParams: Type[] = n
 export function importType(
```
`typeParams` is unused now, could you remove it?

tbosch replied:
Oh, never mind, it is still used.

tbosch commented on modules/@angular/compiler/src/output/ts_emitter.ts:
```diff
@@ -372,10 +392,28 @@ class _TsEmitterVisitor extends AbstractEmitterVisitor implements o.TypeVisitor
+      // If we are in a type expreession that refers to a generic type then supply
+      // the required type parameters. If there were not enough type parameters
+      // supplied, supply any as the type. Outside a type expression the reference
+      // should not supply type parameters and be treated as a simple reference to
```
as a simple _value_ reference

tbosch commented on modules/@angular/compiler/src/output/ts_emitter.ts:
```diff
@@ -372,10 +392,28 @@ class _TsEmitterVisitor extends AbstractEmitterVisitor implements o.TypeVisitor
+      // supplied, supply any as the type. Outside a type expression the reference
+      // should not supply type parameters and be treated as a simple reference to
+      // the constructor function itself.
+      const suppliedParameters = (isPresent(typeParams) && typeParams.length) || 0;
```
drop `isPresent`

tbosch replied:
i.e. just `typeParams && typeParams.length`

tbosch commented on modules/@angular/language-service/src/expressions.ts:
```diff
@@ -753,8 +754,10 @@ function getVarDeclarations(info: TemplateInfo, path: TemplateAstPath): SymbolDe
 function refinedVariableType(
     type: Symbol, info: TemplateInfo, templateElement: EmbeddedTemplateAst): Symbol {
   // Special case the ngFor directive
-  const ngForDirective =
```
Could you move this change into the last commit?

tbosch commented on modules/@angular/language-service/src/expressions.ts:
```diff
@@ -728,7 +728,8 @@ function getVarDeclarations(info: TemplateInfo, path: TemplateAstPath): SymbolDe
           const value = context.get(variable.value);
           if (value) {
             type = value.type;
-            if (info.template.query.getTypeKind(type) === BuiltinType.Any) {
```
Could you move this change into the last commit?

tbosch commented on modules/@angular/language-service/src/types.ts:
```diff
@@ -190,6 +190,11 @@ export enum BuiltinType {
   Null,
 
   /**
+   * the type is an unbound type parameter.
```
Could you move this change into the last commit? Or its not clear to me how this commit requires this change...

tbosch commented on modules/@angular/language-service/src/typescript_host.ts:
```diff
@@ -1263,6 +1263,8 @@ function typeKindOf(type: ts.Type): BuiltinType {
         }
       }
       return candidate;
+    } else if (type.flags & ts.TypeFlags.TypeParameter) {
```
Could you move this change into the last commit? Or its not clear to me how this commit requires this change...

tbosch commented:
Could you add an example to the commit message for the breaking change `A definition of `Iterable<T>` is now required to correctly compile...`?

tbosch commented on modules/@angular/common/src/directives/ng_for_of.ts:
```diff
@@ -102,17 +105,17 @@ export class NgFor implements DoCheck, OnChanges {
 
   @Input()
-  set ngForTemplate(value: TemplateRef<NgForRow>) {
+  set ngForTemplate(value: TemplateRef<NgForOfRow<T>>) {  // TODO: TemplateRef<Partial<NgForRow<T>>>
```
TODO?

tbosch commented:
Can you double check the commit message with @IgorMinar ?

tbosch commented on modules/@angular/core/src/change_detection/differs/iterable_differs.ts:
```diff
@@ -10,6 +10,12 @@ import {Optional, Provider, SkipSelf} from '../../di';
 import {ChangeDetectorRef} from '../change_detector_ref';
 
+/**
+ * A type describing supported interable types.
```
typo: iterable types

tbosch replied:
Why do we need `NgIterable`, and not just a plain `Iterable<T>`? Is it related to the @mhevery's comment at https://github.com/angular/angular/blob/master/modules/@angular/core/src/change_detection/differs/iterable_differs.ts#L29?
If we don't need it, remove it, otherwise please leave a comment as to why we need it...

tbosch commented on modules/@angular/core/src/change_detection/differs/iterable_differs.ts:
```diff
@@ -112,13 +118,19 @@ export interface CollectionChangeRecord<V> extends IterableChangeRecord<V> {}
+ * Nolonger used.
  *
- * @stable
+ * @deprecated v4.0.0 - Use TrackByFunction instead
```
Can you double check if that is our convention now? I.e. adding the version number to @deprecated? If so, please also add the version number to the @deprecated tag on `NgFor`

chuckjaz committed:
feat(compiler): generate type parameters for generic type references (#14104)
